modules/ssl/ssl_engine_pphrase.c

```diff
@@ -600,3 +600,336 @@ int ssl_pphrase_Handle_CB(char *buf, int bufsize, int verify, void *srv)
+{
+    pphrase_cb_arg_t *ppcb_arg;
+    SSLSrvConfigRec *sc;
+    const char *prompt;
+    int i;
+    int bufsize;
+    int len;
+    char *buf;
+
+    ppcb_arg = (pphrase_cb_arg_t *) UI_get0_user_data(ui);
+
+    if (ppcb_arg == NULL) {
+        ap_log_error(APLOG_MARK, APLOG_ERR, 0, NULL, APLOGNO()
+                     "Init: NULL UI callback parameter ");
+        return 0;
+    }
+
+    sc = mySrvConfig(ppcb_arg->s);
+
+    prompt = UI_get0_output_string(uis);
+    if (prompt == NULL) {
+        prompt = "Enter pass phrase:";
+    }
+
+    /*
+     * Get the maximum expected size and allocate the buffer
+     */
+    bufsize = UI_get_result_maxsize(uis);
+    buf = (char *)apr_pcalloc(ppcb_arg->p, bufsize);
+    if (buf == NULL) {
+        ap_log_error(APLOG_MARK, APLOG_ERR, 0, ppcb_arg->s, APLOGNO()
+                     "Init: Failed to allocate password buffer ");
+        return 0;
+    }
+
+    if (sc->server->pphrase_dialog_type == SSL_PPTYPE_BUILTIN
+            || sc->server->pphrase_dialog_type == SSL_PPTYPE_PIPE) {
+        /*
+         * Get the pass phrase through a callback.
+         * Empty input is not accepted.
+         */
+        for (;;) {
+            if (sc->server->pphrase_dialog_type == SSL_PPTYPE_PIPE) {
+                i = pipe_get_passwd_cb(buf, bufsize, "", FALSE);
+            }
+            else { /* sc->server->pphrase_dialog_type == SSL_PPTYPE_BUILTIN */
+                i = EVP_read_pw_string(buf, bufsize, "", FALSE);
+            }
+            if (i != 0) {
+                //OPENSSL_clear_free(buf, bufsize);
+                return 0;
+            }
+            len = strlen(buf);
+            if (len < 1){
+                apr_file_printf(writetty, "Apache:mod_ssl:Error: Pass phrase"
+                                "empty (needs to be at least 1 character).\n");
+                apr_file_puts(prompt, writetty);
+            }
+            else {
+                break;
+            }
+        }
+    }
+    /*
+     * Filter program
+     */
+    else if (sc->server->pphrase_dialog_type == SSL_PPTYPE_FILTER) {
+        const char *cmd = sc->server->pphrase_dialog_path;
+        const char **argv = apr_palloc(ppcb_arg->p, sizeof(char *) * 3);
+        char *result;
+
+        ap_log_error(APLOG_MARK, APLOG_INFO, 0, ppcb_arg->s, APLOGNO()
+                     "Init: Requesting pass phrase from dialog filter "
+                     "program (%s)", cmd);
+
+        argv[0] = cmd;
+        argv[1] = ppcb_arg->key_id;
+        argv[2] = NULL;
+
+        result = ssl_util_readfilter(ppcb_arg->s, ppcb_arg->p, cmd, argv);
+        apr_cpystrn(buf, result, bufsize);
+        len = strlen(buf);
+    }
+
+    /*
+     * Ok, we now have the pass phrase, so give it back
+     */
+    ppcb_arg->cpPassPhraseCur = apr_pstrdup(ppcb_arg->p, buf);
+    UI_set_result(ui, uis, buf);
+
+    /*
+     * Try to overwrite sensitive data
+     */
+    memset(buf, 0x00, bufsize);
```

Comment:
This memset call may be removed by the compiler. You have to use either ``memset_s`` from C11 or ``OPENSSL_cleanse()``.

Reply:
Thank you for pointing this. I know it is not an excuse, but I just kept the way it was written there (apache's mod_ssl uses memset everywhere to __try__ to overwrite sensitive data).
Maybe I should send a new patch fixing this in other parts of their code.